### DISCUSSION TITLE
Updating OpenTTD to use liballegro5

### DISCUSSION THREAD

**Discussion by krysclarke** [OpenTTD/OpenTTD#10096] [Features]
Given that liballegro5 has been out for some time, and naturally when I wanted to start helping to contribute to OpenTTD with some coding work, I noticed that OpenTTD is pinned against liballegro4.
To that end, I raised a bug report, and was informed that this is intentional but that a PR would be accepted.

I'm now starting the work to get OpenTTD to compile against liballegro5.

- Side note: The build of OpenTTD v12.2 that's in Debian Unstable does NOT have allegro support compiled in. Everything works fine.

This discussion post is to let you all know how I'm going, and what I've found that's different.

- V5 has been split into several files, with 'addons' to cover many areas (audio being one).
- V5 does NOT support MIDI ([liballegro github discussion](https://github.com/liballeg/allegro5/issues/444), see last comment) - minor annoyance, but fluidsynth and extmidi are still available, so this will mean that src/music/allegro_m.{cpp,h} aren't needed any more.

##  Sunday, 23 Oct, 2022
- V5 has completely re-arranged the audio stream calls.  Instead of just 4 audio stream related methods, there's now over 30!

## Monday, 24 Oct, 2022
- V5 has also made changes to the video routines.
- In light of the suggestion by nielsmh to add v5 support alongside v4 instead of replacing it, I accept the challenge.
To that end, what are people's thoughts on the method to expose this to users:
1. allowing a mix-n-match approach, allowing users to specify specific versions of allegro (i.e. 'allegro' and 'allegro5' to videodriver & sounddriver config file entries) for sound and video drivers (NOTE the above about v5 NOT supporting MIDI, therefore no need to make it available for musicdriver options); or
2. default to v5, if available, and auto-fallback to v4 (again, for musicdriver, v5 isn't possible due to lack of MIDI support); or
3. Something else???

- hmmm... Getting both v4 & v5 to be detected by cmake: done.  Getting it to compile: not working - BOTH provide an `allegro.h` file, which complicates the setup within cmake.  I have a working find_path for each, but getting find_library to play ball is eluding me so far, but I'll keep on chewing away at this elephant one mouthful at a time.

## Saturday, 29 Oct, 2022
- No progress made on sorting out v4 allegro.h vs v5 allegro5/allegro.h at link time, either - seems that it'll have to be one of the other, from what I've read, as it seems that the linker can't tell which header file is which as they have the same name (excluding the path).

**Comment by nielsmh** [comment]
Consider implementing it as a new, separate set of drivers so Allegro 4 support can be kept around. I'm not sure if anyone's really using it, but I think there is still some support for BeOS/Haiku in OpenTTD.
